finetune/network_modules.py

```diff
@@ -312,3 +312,38 @@ def seq_lab_internal(hidden):
                 'transition_matrix': transition_params
             }
         }
+
+
+def cosine_similarity(hidden1, hidden2, targets, n_targets, dropout_placeholder, config, train=False, reuse=None, **kwargs):
+    """
+    A simple model to compute cosine similarity between two document embeddings.
+
+    :param hidden1: The output of the featurizer for document 1. [batch_size, embed_dim]
+    :param hidden2: The output of the featurizer for document 2. [batch_size, embed_dim]
+    :param targets: One hot encoded target ids. [batch_size, n_classes]
+    :param n_targets: A python int containing the number of classes that the model should be learning to predict over.
+    :param dropout_placeholder:
+    :param config: A config object, containing all parameters for the featurizer.
+    :param train: If this flag is true, dropout and losses are added to the graph.
+    :param reuse: Should reuse be set within this scope.
+    :param kwargs: Spare arguments.
+    :return: dict containing:
+        logits: The unnormalised log probabilities of each class.
+        losses: The loss for the classifier.
+    """
+    with tf.variable_scope('cosine_similarity', reuse=reuse):
+        hidden1 = dropout(hidden1, config.clf_p_drop, train, dropout_placeholder)
+        hidden2 = dropout(hidden2, config.clf_p_drop, train, dropout_placeholder)
+        cos_sim_logits = tf.sigmoid(tf.multiply(hidden1, hidden2))
```

Comment:
I don't think this is the loss function typically used in Siamese networks.  You might want to take a look at the contrastive loss function from the [tensorflow examples](https://github.com/tiagofrepereira2012/examples.tensorflow/blob/master/examples/tensorflow/script/train_mnist_siamese.py) or from this [hackernoon article](https://hackernoon.com/one-shot-learning-with-siamese-networks-in-pytorch-8ddaab10340e).

Reply:
Technically, you're right. Since I had originally intended to use `finetune` for my own research, I had a hack-y way of implementing the contrastive loss with negative samples using binary cross entropy.   Emailing you about this.